### PR TITLE
[Feature]Access check in client

### PR DESCRIPTION
**Pull request by duanmeng**
### What changes were proposed in this pull request?
- Add a delegation shuffle manager
- choose the actual shuffle manager  (sort or rss)

### Why are the changes needed?
This pr is related to #61, which add access client.

### Does this PR introduce _any_ user-facing change?
Yes, client need to config `spark.shuffle.manager` explicitly and add the following config , `spark.rss.accessInfoKey` it specifies which is the config key of the accessInfo and then add the config duration submition.

### How was this patch tested?
UT and system test

### CODE REVIEW

File: internal-client/src/main/java/com/tencent/rss/client/response/RssAccessClusterResponse.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package com.tencent.rss.client.response;
+
+public class RssAccessClusterResponse extends ClientResponse {
+
+  public RssAccessClusterResponse(ResponseStatusCode statusCode, String messge) {
+    super(statusCode, messge);
+  }
+
```
Comment (jerqi) [inline internal-client/src/main/java/com/tencent/rss/client/response/RssAccessClusterResponse.java:26]: too many blank lines.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java [hunk 2]
```diff
@@ -79,4 +79,13 @@
   // has huge amount blockIds, eg, 10B.
   public static String RSS_CLIENT_BITMAP_SPLIT_NUM = "spark.rss.client.bitmap.splitNum";
   public static int RSS_CLIENT_BITMAP_SPLIT_NUM_DEFAULT_VALUE = 1;
+  public static String RSS_ACCESS_ID_KEY = "spark.rss.accessIdKey";
+  public static String RSS_ACCESS_ID_KEY_DEFAULT_VALUE = "";
+  public static String RSS_ACCESS_ID_DEFAULT_VALUE = "";
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:84]: What's the difference between `RSS_ACCESS_ID_KEY` and `RSS_ACCESS_ID_DEFAULT_VALUE`?

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  private final int accessTimeoutMs;
+
+  public DelegationRssShuffleManager(SparkConf sparkConf, boolean isDriver) throws Exception {
+    sortShuffleManagerName = sparkConf.get(
+        RssClientConfig.RSS_SORT_SHUFFLE_MANAGER_IMPL,
+        RssClientConfig.RSS_SORT_SHUFFLE_MANAGER_IMPL_DEFAULT_VALUE);
+    accessTimeoutMs = sparkConf.getInt(
+        RssClientConfig.RSS_ACCESS_TIMEOUT_MS,
+        RssClientConfig.RSS_ACCESS_TIMEOUT_MS_DEFAULT_VALUE);
+    if (isDriver) {
```
Comment (frankliee) [inline client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:55]: In local mode, the driver and executor will live together.

File: proto/src/main/proto/Rss.proto [hunk 4]
```diff
@@ -231,6 +231,7 @@ enum StatusCode {
   NO_PARTITION = 5;
   INTERNAL_ERROR = 6;
   TIMEOUT = 7;
+  ACCESS_FAIL = 8;
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply proto/src/main/proto/Rss.proto]: Yes, will rebase once that pr merged.

File: internal-client/src/main/java/com/tencent/rss/client/response/RssAccessClusterResponse.java (hunk 1 above)
Comment (duanmeng) [reply internal-client/src/main/java/com/tencent/rss/client/response/RssAccessClusterResponse.java]: Fixed

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      coordinatorClients = Lists.newArrayList();
+      delegate = createShuffleManagerInExecutor();
+    }
+
+    if (delegate == null) {
+      throw new RssException("Fail to create shuffle manager!");
+    }
+  }
+
+  private ShuffleManager createShuffleManagerInDriver() throws RssException {
```
Comment (colinmjj) [inline client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:73]: why need both createShuffleManagerInDriver and createShuffleManagerInExecutor?
it should be merged in one method

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java (hunk 2 above)
Comment (duanmeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:83]: This is kind of specific, I will remote it.

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java (hunk 3 above)
Comment (duanmeng) [reply client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:57]: This is a parameter of RssShuffleManager

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java (hunk 5 above)
Comment (duanmeng) [reply client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:70]: These two functions wrapper the code in the if else block of the constructor. They follow different procedure, driver will send request to decide use rss or not, create related shuffle manager and set the useRss flag in spark conf, executor just check the useRss flag to decide which shuffle manager to use and create it.

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    return shuffleManager;
+  }
+
+  public ShuffleManager getDelegate() {
+    return delegate;
+  }
+
+  @Override
+  public <K, V, C> ShuffleHandle registerShuffle(int shuffleId, int numMaps, ShuffleDependency<K, V, C> dependency) {
```
Comment (jerqi) [inline client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:154]: @frankliee Could you see the compatibility of 3.0 and 2.3?

**Comment by jerqi** [comment]
Should we add an integration test case?

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+      }
+    }
+
+    return false;
+  }
+
+  private ShuffleManager createShuffleManagerInExecutor() throws RssException {
+    ShuffleManager shuffleManager;
+    // get useRSS from spark conf
+    boolean useRSS = Boolean.parseBoolean(sparkConf.get(
```
Comment (colinmjj) [inline client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:131]: sparkConf.getBoolean

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java [hunk 8]
```diff
@@ -79,4 +79,12 @@
   // has huge amount blockIds, eg, 10B.
   public static String RSS_CLIENT_BITMAP_SPLIT_NUM = "spark.rss.client.bitmap.splitNum";
   public static int RSS_CLIENT_BITMAP_SPLIT_NUM_DEFAULT_VALUE = 1;
+  public static String RSS_ACCESS_ID = "spark.rss.access.id";
+  public static String RSS_ACCESS_ID_DEFAULT_VALUE = "";
```
Comment (colinmjj) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:83]: remove empty default value

File: coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java [hunk 9]
```diff
@@ -59,12 +59,12 @@ public AccessCandidatesChecker(AccessManager accessManager) {
...
     if (!candidates.get().contains(accessId)) {
-      String msg = String.format("Reject accessInfo[%s] accessId[%s] access request.", accessInfo, accessId);
-      LOG.debug(msg);
+      String msg = String.format("Denied by AccessCandidatesChecker, accessInfo[%s].", accessInfo);
+      LOG.debug("Candidates is {}, {}", candidates.get(), msg);
       return new AccessCheckResult(false, msg);
     }
 
-    return new AccessCheckResult(true, "");
+    return new AccessCheckResult(true, "SUCCESS");
```
Comment (colinmjj) [inline coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java:67]: make SUCCESS as a constant

File: coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java [hunk 10]
```diff
@@ -66,7 +66,8 @@
   public static final ConfigOption<String> COORDINATOR_ACCESS_CHECKERS = ConfigOptions
       .key("rss.coordinator.access.checkers")
       .stringType()
-      .noDefaultValue()
+      .defaultValue("com.tencent.rss.coordinator.AccessCandidatesChecker,"
```
Comment (colinmjj) [inline coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java:69]: user should config the checkers by themselves, or there will be unexpected actions for them

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java (hunk 6 above)
Comment (frankliee) [reply client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:154]: mvn package -Pspark2.3
mvn package -Pspark3.0

File: coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java (hunk 10 above)
Comment (duanmeng) [reply coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java:69]: this is the coordinator's configuration

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java (hunk 8 above)
Comment (duanmeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java]: OK

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java (hunk 7 above)
Comment (duanmeng) [reply client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java]: ok

File: coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java (hunk 9 above)
Comment (duanmeng) [reply coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java:67]: ok

**Comment by duanmeng** [comment]
> Should we add an integration test case?

It's in integration-test/common named AccessClusterTest

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java [hunk 11]
```diff
@@ -79,4 +79,11 @@
   // has huge amount blockIds, eg, 10B.
   public static String RSS_CLIENT_BITMAP_SPLIT_NUM = "spark.rss.client.bitmap.splitNum";
   public static int RSS_CLIENT_BITMAP_SPLIT_NUM_DEFAULT_VALUE = 1;
+  public static String RSS_ACCESS_ID = "spark.rss.access.id";
+  public static String RSS_ACCESS_TIMEOUT_MS = "spark.rss.access.timeout.ms";
+  public static int RSS_ACCESS_TIMEOUT_MS_DEFAULT_VALUE = 10000;
+  public static String RSS_USE_RSS_SHUFFLE_MANAGER = "spark.rss.useRssShuffleManager";
+  public static boolean RSS_USE_RSS_SHUFFLE_MANAGER_DEFAULT_VALUE = false;
+  public static String RSS_SORT_SHUFFLE_MANAGER_IMPL = "spark.rss.sort.shuffle.manager.impl";
```
Comment (colinmjj) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:87]: The parameter is unnecessary, make org.apache.spark.shuffle.sort.SortShuffleManager as a constant.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java [hunk 12]
```diff
@@ -79,4 +79,11 @@
   // has huge amount blockIds, eg, 10B.
   public static String RSS_CLIENT_BITMAP_SPLIT_NUM = "spark.rss.client.bitmap.splitNum";
   public static int RSS_CLIENT_BITMAP_SPLIT_NUM_DEFAULT_VALUE = 1;
+  public static String RSS_ACCESS_ID = "spark.rss.access.id";
+  public static String RSS_ACCESS_TIMEOUT_MS = "spark.rss.access.timeout.ms";
+  public static int RSS_ACCESS_TIMEOUT_MS_DEFAULT_VALUE = 10000;
+  public static String RSS_USE_RSS_SHUFFLE_MANAGER = "spark.rss.useRssShuffleManager";
```
Comment (colinmjj) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:85]: spark.rss.useRssShuffleManager -> spark.rss.enabled

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+  public DelegationRssShuffleManager(SparkConf sparkConf, boolean isDriver) throws Exception {
+    this.sparkConf = sparkConf;
+    sortShuffleManagerName = sparkConf.get(
+        RssClientConfig.RSS_SORT_SHUFFLE_MANAGER_IMPL,
+        RssClientConfig.RSS_SORT_SHUFFLE_MANAGER_IMPL_DEFAULT_VALUE);
+    accessTimeoutMs = sparkConf.getInt(
+        RssClientConfig.RSS_ACCESS_TIMEOUT_MS,
+        RssClientConfig.RSS_ACCESS_TIMEOUT_MS_DEFAULT_VALUE);
+    if (isDriver) {
+      coordinatorClients = RssShuffleUtils.createCoordinatorClients(sparkConf);
```
Comment (colinmjj) [inline client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:58]: if shuffle with RssShuffleManager, DelegationRssShuffleManager & RssShuffleManager will have different coordinator clients. These clients should be shared.

File: coordinator/src/main/java/com/tencent/rss/coordinator/AccessManager.java [hunk 14]
```diff
@@ -62,7 +62,7 @@ public AccessCheckResult handleAccessRequest(AccessInfo accessInfo) {
       }
     }
 
-    return new AccessCheckResult(true, "");
+    return new AccessCheckResult(true, "SUCCESS");
```
Comment (colinmjj) [inline coordinator/src/main/java/com/tencent/rss/coordinator/AccessManager.java:65]: constant?

File: coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java (hunk 9 above)
Comment (colinmjj) [reply coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java:67]: constant?

File: coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java (hunk 10 above)
Comment (colinmjj) [reply coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java:69]: yes, user should config such feature by themselves including client, coordinator, shuffle server

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java (hunk 11 above)
Comment (duanmeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:87]: This param gives users the ability to specify their own shuffle manager.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java (hunk 12 above)
Comment (duanmeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:85]: ok

File: coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java (hunk 10 above)
Comment (duanmeng) [reply coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java:69]: checkers should not be exposed to users, it's kind of administration parameter

File: coordinator/src/main/java/com/tencent/rss/coordinator/AccessManager.java (hunk 14 above)
Comment (duanmeng) [reply coordinator/src/main/java/com/tencent/rss/coordinator/AccessManager.java:65]: ok

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java (hunk 13 above)
Comment (duanmeng) [reply client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:58]: Add a set api in RssShuffleManager and inject the coordinator clients?

File: coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java (hunk 9 above)
Comment (duanmeng) [reply coordinator/src/main/java/com/tencent/rss/coordinator/AccessCandidatesChecker.java:67]: ok add in Constants class

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java (hunk 13 above)
Comment (duanmeng) [reply client-spark/spark2/src/main/java/org/apache/spark/shuffle/DelegationRssShuffleManager.java:58]: As discussed offline, will add refactor the coordinator clients creation in next pr.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java (hunk 11 above)
Comment (colinmjj) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:87]: It's unnecessary

File: coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java (hunk 10 above)
Comment (duanmeng) [reply coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorConf.java]: will only make load checker be default

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java (hunk 11 above)
Comment (duanmeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:87]: removed

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java [hunk 15]
```diff
@@ -79,4 +79,9 @@
   // has huge amount blockIds, eg, 10B.
   public static String RSS_CLIENT_BITMAP_SPLIT_NUM = "spark.rss.client.bitmap.splitNum";
   public static int RSS_CLIENT_BITMAP_SPLIT_NUM_DEFAULT_VALUE = 1;
+  public static String RSS_ACCESS_ID = "spark.rss.access.id";
+  public static String RSS_ACCESS_TIMEOUT_MS = "spark.rss.access.timeout.ms";
+  public static int RSS_ACCESS_TIMEOUT_MS_DEFAULT_VALUE = 10000;
+  public static String RSS_USE_RSS_SHUFFLE_MANAGER = "spark.rss.enabled";
```
Comment (colinmjj) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java:85]: RSS_USE_RSS_SHUFFLE_MANAGER -> rss_enabled

Comment (duanmeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssClientConfig.java]: Fixed

**Review by colinmjj — APPROVED** [review]
LGTM

**Pull request merged**